### ISSUE TITLE
Lag with Milights

### ISSUE THREAD

**Issue by jorisvanr** [mariusmotea/diyHue#382]
I experience a lot of lag with my Milights and the Hue app and i'm not sure what causes it. At first I thought it was a range issue but the lights in a different room are actually almost performing better than the ones in the room with the transmitter. 

When I select a scene in the app, some of the lights change to the designated color but almost always on or 2 lights don't change at all. I have 12 Milights in total. 6 of which are on 1 light switch and therefore on 1 Hub ID. 

The WS2812B ledstrip works every time so it really is the Milights (or the transmitter) and not the bridge. 

A lot of the lights are displayed as "unreachable" in the app on and off as well.  Maybe this has to do with it? 

Edit: Controlling individual lights often works fine (on / off / colour changes) as long as I don't do the commands to fast. Changing the scene gives some problems. Most of my lights are added to the same group during the connection process as I did not think this would be an issue as i'd make groups in the app anyways. 

ps. I'm not sure if this is the right way of posting the bugs. I don't want to come across impatient or anything. I think this project is awesome and it is what i've been looking for for a long time. Keep it up 👍

**Comment by mariusmotea** [comment]
It is because the milight hub is quickly overloaded by http requests sent at every 0.1 seconds. I will push a new update to fix this.

**Comment by cooltom** [comment]
Hi, any way to add option to comunicate over MQTT connection to ESP Milight hub into your Script? :)

https://github.com/sidoh/esp8266_milight_hub/issues/343#issuecomment-424624781

**Comment by mariusmotea** [comment]
Can you please test this:

Open HueEmulator3.py script from /opt/hue-emulator and replace following functions

`Thread(target=sendLightRequest, args=[light, bridge_config["scenes"][put_dictionary["scene"]]["lightstates"][light]]).start()`
with 
`sendLightRequest(light, bridge_config["scenes"][put_dictionary["scene"]]["lightstates"][light]])`

`Thread(target=sendLightRequest, args=[light, put_dictionary]).start()`
with
`sendLightRequest(light, put_dictionary)`

`Thread(target=sendLightRequest,args=[url_pices[4], put_dictionary]).start()`
with
`sendLightRequest(url_pices[4], put_dictionary)`

basically you need to apply this to everything that starts with `Thread(target=sendLightRequest` and then restart hue-emulator service.

**Comment by mariusmotea** [comment]
Hi,

I made a new commit in develop branch that will hangle the milight request one by one to not overload the milight bridge. If your main lights are milight i recomment to upgrade to this version.

https://github.com/mariusmotea/diyHue/commit/fd8a330f7ea94e70a3f59ea9165239d985232a1a

**Comment by mariusmotea** [comment]
That commit is a total rework of how lights are handled. Just download HueEmulator3.py from develop branch and replace your.

**Comment by cooltom** [comment]
getting error, when run updated script ..

[23:39:34] openhabian@openHABianPi:/opt/hue-emulator$ sudo systemctl stop hue-emulator.service
[23:40:01] openhabian@openHABianPi:/opt/hue-emulator$ sudo ./HueEmulator3.py
Traceback (most recent call last):
  File "./HueEmulator3.py", line 24, in <module>
    from functions import light_types, nextFreeId
ImportError: cannot import name 'light_types'
[23:40:06] openhabian@openHABianPi:/opt/hue-emulator$

**Comment by mariusmotea** [comment]
I believe you use an old version that don't have the new files from functions directory. Please downlaod them from master branch located here: https://github.com/mariusmotea/diyHue/tree/master/BridgeEmulator/functions

**Comment by cooltom** [comment]
Oh true, working now. Getting this error message, but all lights work and can be controled. Something stucked in config ?

ERROR:root:light 5 is unreachable
Traceback (most recent call last):
  File "./HueEmulator3.py", line 731, in syncWithLights
    bridge_config["lights"][light]["state"]["xy"] = convert_rgb_xy(light_data["color"]["r"], light_data["color"]["g"], light_data["color"]["b"])
KeyError: 'color'
ERROR:root:light 2 is unreachable
Traceback (most recent call last):
  File "./HueEmulator3.py", line 731, in syncWithLights
    bridge_config["lights"][light]["state"]["xy"] = convert_rgb_xy(light_data["color"]["r"], light_data["color"]["g"], light_data["color"]["b"])
KeyError: 'color'
ERROR:root:light 3 is unreachable
Traceback (most recent call last):
  File "./HueEmulator3.py", line 731, in syncWithLights
    bridge_config["lights"][light]["state"]["xy"] = convert_rgb_xy(light_data["color"]["r"], light_data["color"]["g"], light_data["color"]["b"])
KeyError: 'color'
ERROR:root:light 1 is unreachable
Traceback (most recent call last):
  File "./HueEmulator3.py", line 731, in syncWithLights
    bridge_config["lights"][light]["state"]["xy"] = convert_rgb_xy(light_data["color"]["r"], light_data["color"]["g"], light_data["color"]["b"])
KeyError: 'color'
ERROR:root:light 7 is unreachable
Traceback (most recent call last):
  File "./HueEmulator3.py", line 731, in syncWithLights
    bridge_config["lights"][light]["state"]["xy"] = convert_rgb_xy(light_data["color"]["r"], light_data["color"]["g"], light_data["color"]["b"])
KeyError: 'color'
ERROR:root:light 6 is unreachable
Traceback (most recent call last):
  File "./HueEmulator3.py", line 731, in syncWithLights
    bridge_config["lights"][light]["state"]["xy"] = convert_rgb_xy(light_data["color"]["r"], light_data["color"]["g"], light_data["color"]["b"])
KeyError: 'color'

**Comment by mariusmotea** [comment]
Possible change in Milight hub API because the color key must be present according the documentation

> 1. color. Hash containing RGB color. All keys for r, g, and b should be present. For example, {"r":255,"g":200,"b":255}.

**Comment by mariusmotea** [comment]
You can check this with following command:

`curl http://{MILIGHT HUB IP}/json.htm?type=devices&rid={REMOTE ID}`

and provide here the output.

**Comment by cooltom** [comment]
True is that I downgrade ESPMilight hub firmware to 1.8.1 as it seems like have faster response ...

if RID is decimal bridge ID, than I got this response

`[1] 3416`

**Comment by mariusmotea** [comment]
do this:

curl "http://{MILIGHT HUB IP}/json.htm?type=devices&rid={REMOTE ID}"

Because "&" is interpreted by bash if is not inside "" and the command run in background  incomplete.

**Comment by cooltom** [comment]
`http://IP/gateways/5547/rgb_cct/1`
returned
`{"state":"OFF","level":0,"hue":261,"saturation":100,"bulb_mode":"color"}`

**Comment by cooltom** [comment]
so i dont know .. on the top ESP commands receave over MQTT, bottom DiyHue sent and left ping to ESP

http://czechmix.cz/DSC_4575.JPG

diyHue always send set of command, ESP get stoned, than execute few of them and get stoned again. Tested with Hue Disco.

And after update Hue app can not connect with Hue App to DiyHue anymore, Clicking on activate and Hue app endup wtih button not pressed connection error

WEB - - [02/Oct/2018 01:48:26] "GET /hue/linkbutton?username=&password=&action=Activate HTTP/1.1" 200 -
MOBILE IP - - [02/Oct/2018 01:48:33] "GET /api/nouser/config HTTP/1.1" 200 -
MOBILE IP - - [02/Oct/2018 01:48:44] "GET /api/nouser/config HTTP/1.1" 200 -
WEB - - [02/Oct/2018 01:48:46] "GET /hue/linkbutton?username=&password=&action=Activate HTTP/1.1" 200 -
MOBILE IP - - [02/Oct/2018 01:49:04] "GET /api/nouser/config HTTP/1.1" 200 -
MOBILE IP - - [02/Oct/2018 01:49:06] "GET /api/nouser/config HTTP/1.1" 200 -
WEB - - [02/Oct/2018 01:49:12] "GET /hue/linkbutton?username=&password=&action=Activate HTTP/1.1" 200 -
WEB - - [02/Oct/2018 01:49:15] "GET /hue/linkbutton?username=&password=&action=Activate HTTP/1.1" 200 -
WEB - - [02/Oct/2018 01:49:19] "GET /hue/linkbutton?username=&password=&action=Activate HTTP/1.1" 200 -
WEB - - [02/Oct/2018 01:49:26] "GET /hue/linkbutton?username=&password=&action=Activate HTTP/1.1" 200 -
MOBILE IP - - [02/Oct/2018 01:49:26] "GET /api/nouser/config HTTP/1.1" 200 -
MOBILE IP - - [02/Oct/2018 01:49:46] "GET /api/nouser/config HTTP/1.1" 200 -
MOBILE IP - - [02/Oct/2018 01:50:06] "GET /api/nouser/config HTTP/1.1" 200 -
MOBILE IP - - [02/Oct/2018 01:50:26] "GET /api/nouser/config HTTP/1.1" 200 -

**Comment by jorisvanr** [comment]
I just updated my system. It seems to work a little bit better / faster now but I still have issues with lights not responding to the request. 

For example I have 6 GU10 lights that are all in 1 group. Very often 1 or 2 lights don't follow the rest. This means 1 light will stay on when the rest is turned off. Could this be caused by the hardware or the software? I have less issues with 2 lights that are in another room so I don't think it is a connection problem. 

Some of my other lights also sometimes still don't respond. The status used to update in the app (so if they did not turn off I'd see it in the app as well and could turn it off manually anyways) but now it just says it is off even if it isn't.

**Comment by cooltom** [comment]
Do u use official Hue app? If yes, what version of Hue app?

**Comment by jorisvanr** [comment]
Yes I do. V. 3.4.0 (9965) on iOs

**Comment by cooltom** [comment]
> 
> 
> Yes I do. V. 3.4.0 (9965) on iOs

on android but same number and can not connect to bridge :( Old version Hue Disco (had bridge already paired) can connect ...

**Comment by cheesemarathon** [comment]
Ok, this issue seems to have gone a bit crazy and so i'm not sure if it's been fixed!

@jorisvanr Are you still having issues with lag? If so would you please explain them again so we can try fix this?

@cooltom are you having lagging issues also? If so, could you please explain your issue again so we can try fix it? You also said:

> Hi, any way to add option to comunicate over MQTT connection to ESP Milight hub into your Script? :)

Is this a feature you want to request or was it a suggestion to fix the lag? If its a feature request, please open a new issue for it.

**Comment by jorisvanr** [comment]
Haha yes we got a little off-topic here. 

Even tough performance is improved a bit I still experience a lot of bugs where all lights are changed simultaneously. So when I set a scene in the hue app or tell Alexa to turn of all lights, some of the lights don’t respond. The Mi-Lights often show up as unreachable in the app as well. When I turn them on and off in the app they do respond afterwards. 

Btw besides this I also did a fix where [the esp8266 and the transmitter were causing interference](https://github.com/sidoh/esp8266_milight_hub/issues/296#issuecomment-429399608). This helped where not all lights of the same group responded at the same time.

**Comment by mariusmotea** [comment]
Ok, i will push a new update with a 50ms delay between the requests, maybe this will be enough for milight hub to process all the requests successfuly.

**Comment by cooltom** [comment]
Are you able connect from last Hue App v 3.4.0 (6280) Android. It see bridge. But can not see that u press button (click on activate). Seems like all new versions have some diferent way how to autorize/pair ...

**Comment by cheesemarathon** [comment]
@cooltom @jorisvanr Can you test @mariusmotea's commit ee74981 that should fix the milight hub being unable to process the requests fast enough? Just replace your HueEmulator3.py with the latest one from GitHub.

**Comment by cooltom** [comment]
> 
> 
> @cooltom @jorisvanr Can you test @mariusmotea's commit [ee74981](https://github.com/mariusmotea/diyHue/commit/ee74981135b1db3caa4773cce674b640b861561d) that should fix the milight hub being unable to process the requests fast enough? Just replace your HueEmulator3.py with the latest one from GitHub.

Still same, maybe bit better. Once i turn on some lightapp (Old Hue Disco) ESP start diconnecting and connecting from wifi and not response for ping. Once i stop it it go back to normal.
If I change changes to 10s in Hue Disco app, some ping loose, but at least bulbs react sometimes to app changes.
So I should play with this delay number?

I will create new issue for new versions of Hue Apps (seems some update or mistake on my site) as only old versions is working - can not test old hue app again as I updated it already :(

for MQTT connection I create new feature request.

**Comment by mariusmotea** [comment]
@cooltom did you try to set the esp clock to 160MHz? Also now you can manually set the delay between requests, even if you increase that, the milight hub still hang?

**Comment by cooltom** [comment]
> 
> 
> @cooltom did you try to set the esp clock to 160MHz? Also now you can manually set the delay between requests, even if you increase that, the milight hub still hang?

Hmm I didnt. Its somewhere in firmware??

**Comment by mariusmotea** [comment]
Look in [this picture](https://iot-playground.com/images/articles/033/24-9-2015_23-33-32.png), the cpu clock is 80Mhz, but you can select also 160Mhz

**Comment by cooltom** [comment]
> 
> 
> Look in [this picture](https://iot-playground.com/images/articles/033/24-9-2015_23-33-32.png), the cpu clock is 80Mhz, but you can select also 160Mhz

I see using programing App. Can is survive double FRQ? :)

**Comment by mariusmotea** [comment]
I test it in one esp light and i had no issues. For shure the chip will consume more power in this mode.

**Comment by cheesemarathon** [comment]
@cooltom Did the overclocking the ESP help?

**Comment by cheesemarathon** [comment]
I think the issue here has been clearly identified. WiFi interference and the low processing power of the ESP8266 bridge. Marius has made some changes to mitigate these issues and overclocking the ESP can help. Without rewriting the bridge software to run on a more powerful system, I think we have done all we can.
